src/ComputeSharp.SourceGeneration/Extensions/ISymbolExtensions.cs

```diff
@@ -155,4 +156,24 @@ public static bool TryGetSyntaxNode<T>(this ISymbol symbol, CancellationToken to
 
         return candidateNode is not null;
     }
+
+    /// <summary>
+    /// Gets the first symbol of a specific type among the ancestors of a given symbol.
+    /// </summary>
+    /// <param name="symbol">The input symbol to start the search from.</param>
+    /// <param name="predicate">An optional predicate to filter symbols.</param>
+    /// <remarks>The resulting symbol, if a match was found.</remarks>
+    public static TSymbol? FirstAncestorOrSelf<TSymbol>(this ISymbol symbol, Func<TSymbol, bool>? predicate = null)
+        where TSymbol : class, ISymbol
+    {
+        for (ISymbol? parentSymbol = symbol; parentSymbol is not null; parentSymbol = parentSymbol.ContainingSymbol)
+        {
+            if (parentSymbol is TSymbol targetSymbol && (predicate?.Invoke(targetSymbol) is not false))
+            {
+                return targetSymbol;
+            }
+        }
+
+        return null;
```

Comment:
shouldn't this return `symbol` ? "First Ancesor _or self_". Either that or the name of this method is clumsy.

Reply:
I'm already returning self if it matches, the `for` loop starts from `symbol` in the first iteration 🙂